SC-419: How to for RBAC (#326)
gobeyondidentity/developer-docs: docs/how-to/configure-rbac.mdx

NOTE: RBAC only comes into play with access management of access tokens of _identities_.
This means that your application's authentication flow must go through the **authorization code** grant type and the access token minted must be on behalf of an _identity_.

An example scenario where RBAC is used is say that you're a developer for an e-commerce site that sells shoes and you have a loyalty program for your customers that contains
two tiers - Free Members and VIP Members. VIP members get exclusive access to new shoes first as well as all features that Free members do. When an end user is attempting
to access a certain part of your site, you can use RBAC to allow or deny access.

Now, let's identify the different components that make up RBAC and walk through the example above.

## Resource Servers and Scopes

Resource servers represent an entity hosting protected resources that the client is attempting to access. This is the entire set of permissions that are allowed on a resource.
These permissions are expressed through scopes, which are strings that represent the set of permissions.

> For our example above, we would create a resource server called Shoes and the scopes that are available could be `shoes:read`, `shoes:buy`, `clearance:buy`, `new-arrivals:read`, `new-arrivals:buy`, etc.

Reviewer: We should probably link to the "Add a resource server" doc.

## Roles

Roles map to a set of allowed scopes. You then assign identities to these roles.

For our example above, we would create the following roles and their scopes:

- Free Member: `shoes:read`, `shoes:buy`, `clearance:buy`
- VIP Member: `shoes:read`, `shoes:buy`, `clearance:buy`, `new-arrivals:read`, `new-arrivals:buy`

Roles come into play in two spots: (1) Token Minting and (2) Token Introspection.

## Token Minting

Whenever an access token is minted on behalf of an identity, the scopes on that token is the intersection of the three following scope sets:

1. Scopes on the `/authorize` request (during the OAuth2 / OIDC flow).
1. Scopes supported by the Application the identity is authenticating into.
1. Scopes tied to the identity's roles.

For our example above, a free member authenticates into the website would have an access token with the scopes `shoes:read`, `shoes:buy`, `clearance:buy`, and a VIP
member's access token would contain the scopes `shoes:read`, `shoes:buy`, `clearance:buy`, `new-arrivals:read`, `new-arrivals:buy`.

## Token Introspection

> Whenever a [token introspection](<link to token introspection>) is called for an access token with an identity as the principal, the scopes returned in the introspect

Reviewer 2: need link here :)

> Whenever a [token introspection](<link to token introspection>) is called for an access token with an identity as the principal, the scopes returned in the introspect
> response will contain **only a subset of the scopes for which the identity has the role(s) for**.

Reviewer 2: ```suggestion
When invoking [token introspection](<link to token introspection>) for an access token associated with an identity as the principal, the returned scopes in the introspection response will contain **only a subset of the scopes for which the identity has the role(s) for**.
```

For the most part, the scopes in the token introspection response will match the scopes on the token itself unless permissions change. But should a user's roles change,
the introspect response will reflect those changes.

For our example above, say a VIP member logs into the website and halfway through their session, they lose their VIP status. Now when the (now free) member attempts to view
the new arrivals, the introspect response should NOT return the `new-arrivals:read` scope on the response and the user should be denied access to that page.

## Putting it all together

> After creating the aforementioned resource server and roles above, we would create an application with the authorization code flow and integrate that with your application.

Reviewer 2: Is there anything we should point out in how to integrate that with the application?

Author: Hmmm I'm not sure if there is anything besides this step. The developer must take steps to introspect within their own application after this.

Now, whenever a user logs in and attempts to access the various parts of your website, RBAC will ensure that the logged in user can only access the resources that are allowed
to them by their role.

> RBAC is an authorization tool that can be used to validate access permissions on the end user, and is just one of the tools that Beyond Identity can help you accomplish your

Reviewer 2: Is there a logical next place to link a user for more tools to accomplish identity and access management goals?

Author: I don't think so, this is it!

> After creating the aforementioned resource server and roles above, we would create an application with the authorization code flow and integrate that with your application.
> Now, whenever a user logs in and attempts to access the various parts of your website, RBAC will ensure that the logged in user can only access the resources that are allowed
> to them by their role.
> 
> RBAC is an authorization tool that can be used to validate access permissions on the end user, and is just one of the tools that Beyond Identity can help you accomplish your
> identity and access management goals.

Reviewer 2: ```suggestion
After creating the aforementioned resource server and roles above, the next step involves the creation of an application employing the authorization code flow using the roles and scopes created. Consequently, whenever a user logs in and attempts to access different sections of your website, Role-Based Access Control (RBAC) will ensure that the logged in user can only access the resources that are allowed to them by their designated role.

RBAC is an authorization tool that can be used to validate access permissions on the end user, and is just one of the tools that Beyond Identity can help you accomplish your identity and access management goals.
```
